pkg/components/rook/component.go

```diff
@@ -29,9 +29,9 @@ func init() {
 }
 
 type component struct {
-	Namespace                string              `hcl:"namespace,optional"`
-	NodeSelectors            []util.NodeSelector `hcl:"node_selector,block"`
```

Comment:
If `util.NodeSelector` is really being used for `nodeAffinity`, maybe we can rename it and add new `NodeSelector` struct? Given that `nodeSelector` is a very common thing for components (or at least I would expect every component to support it eventually), I think it would be really nice to use some structured type for it, even if it's `map[string]string` underneath, as it allows you to implement both validation method and rendering helpers, which will simplify the templating.

Reply:
> If `util.NodeSelector` is really being used for `nodeAffinity`, maybe we can rename it and add new `NodeSelector` struct?

What is the purpose of adding something that is not useful underneath? The chart does not expose any `NodeAffinity` then what is the use of exposing it to the user?

> Given that `nodeSelector` is a very common thing for components (or at least I would expect every component to support it eventually), I think it would be really nice to use some structured type for it, even if it's `map[string]string` underneath, as it allows you to implement both validation method and rendering helpers, which will simplify the templating.

What kind of validation would you like to put into place?

---

We use `nodeSelector` as it is intended to be used as key value pair. 

```yaml
nodeSelector:
  {{- range $key, $value := .NodeSelectors }}
  {{ $key }}: "{{ $value }}"
  {{- end }}
{{- end }}
```

But some fields demand for `nodeAffinity`, so the function `convertNodeSelector` that converts the map of `nodeSelector` to the string. The string format here is specific to this helm chart. This is unconventional string format.

```
provisionerNodeAffinity: key1=value1,value2; key2=value3
```

Widely used string conversion of map is of the format:

```
selector: key1=value1,key2=value2
```

Reply:
>But some fields demand for nodeAffinity, so the function `convertNodeSelector` that converts the map of nodeSelector to the string. The string format here is specific to this helm chart. This is unconventional string format.

I didn't know this is only specific to this chart. Having a function here make sense then.

I looked again into all the variants we have here and it still seems to me, that what I propose is right. We are dealing here with 3 types of data:
- node selector https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#nodeselector
- node affinity https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity
- toleration https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/#concepts

In this particular chart, we use `NodeSelector` and convert it to nodeAffinity, which is then again converted to chart-specific format. So I would propose following patch:
```diff
-type NodeSelector struct {
+type NodeSelector map[string]string
+
+func (ns NodeSelector) Render() (string, error) {
+       if len(ns) == 0 {
+               return "", nil
+       }
+
+       b, err := json.Marshal(ns)
+       if err != nil {
+               return "", err
+       }
+
+       return string(b), nil
+}
+
+func (ns NodeSelector) ToNodeAffinities() []NodeAffinity {
+       ...
+}
+
+type NodeAffinity struct {
```

So, we rename our `NodeSelector` to `NodeAffinity`, as this is what it should be in the first place (this requires change in other components too).

And then we introduce new `NodeSelector` type, which is just an alias type to `map[string]string`, but allows us to use named functions of it, so it's functionality is more object-oriented, so it's easier to read the code. This allows us to have typed methods as `Render()`, which converts `NodeSelector` to YAML snippet to include in the values, as well as `ToNodeAffinities()` method, which does exactly how it's named.

Lastly, I would rename `convertNodeSelector` to something like `flattenNodeSelector` and make it take `NodeSelector` as an argument. This function should also be documented, that it converts to rook-specific format.

Reply:
I see now changes in https://github.com/kinvolk/lokomotive/pull/386, which seems to be going in the right direction.

Reply:
Done changes.